compiler/rustc_hir_analysis/src/check/compare_impl_item.rs

```diff
@@ -290,14 +289,41 @@ fn compare_method_predicate_entailment<'tcx>(
     }
 
     if check_implied_wf == CheckImpliedWfMode::Check {
-        // We need to check that the impl's args are well-formed given
-        // the hybrid param-env (impl + trait method where-clauses).
-        ocx.register_obligation(traits::Obligation::new(
-            infcx.tcx,
-            ObligationCause::dummy(),
-            param_env,
-            ty::Binder::dummy(ty::PredicateKind::WellFormed(unnormalized_impl_fty.into())),
-        ));
+        // See #108544. Annoying, we can end up in cases where, because of winnowing,
+        // we pick param env candidates over a more general impl, leading to more
+        // stricter lifetime requirements than we would otherwise need. This can
+        // trigger the lint. Instead, let's only consider type outlives and
+        // region outlives obligations.
```

Comment:
opened #109491 to experiment with that approach

Reply:
I don't think this is unsound (though, arguably, imperfect).

We check that everything is well formed elsewhere, so there's no unsoundness related to ill-formed types.

The only implied bounds we currently have are those involving regions (or super trait bounds, but those should also be checked elsewhere).

Therefore, we don't necessarily have to care about non-region predicates *for this lint*.

That being said, while in theory, skipping "trivial clauses" is arguably a "nicer" solution, the implementation in #109491 is not *simple*. *And* I think that approach may also have issues when the param env predicates introduce ambiguity that doesn't exist in an empty environment. Though, I'm not sure if this would be caught by other checks.

Reply:
> I don't think this is unsound (though, arguably, imperfect).
> 
> We check that everything is well formed elsewhere, so there's no unsoundness related to ill-formed types.
> 
> The only implied bounds we currently have are those involving regions (or super trait bounds, but those should also be checked elsewhere).
> 
> Therefore, we don't necessarily have to care about non-region predicates for this lint.

We do care about them if these other predicates end up causing us to add additional outlives predicates. Considering that we end up computing `Projection` obligations inside of `compute_implied_outlives_bounds`, this can be the case.

https://github.com/rust-lang/rust/blob/8a7ca936e61d04399198911ee2b07ac110bf17b0/compiler/rustc_traits/src/implied_outlives_bounds.rs#L86-L94

> That being said, while in theory, skipping "trivial clauses" is arguably a "nicer" solution, the implementation in https://github.com/rust-lang/rust/pull/109491 is not simple.

I disagree, the implementation in #109491 is (after enabling the use of the canonical solver mode to get non-fatal overflow) simple. We're checking whether a goal holds in an empty environment. While this needs more lines of code, the behavior of these is clear and also relied on in other areas of the compiler. That approach is also far more obvious to be sound as we don't drop requirements, we only drop assumptions.

> And I think that approach may also have issues when the param env predicates introduce ambiguity that doesn't exist in an empty environment. Though, I'm not sure if this would be caught by other checks.

You mean that the approach in #109491 does not completely fix the issue of incorrect ambiguity/selecting the wrong candidate when proving the wf bounds? That is true and something that can be incrementally improved if people encounter that issue .

I very strongly prefer trivially sound (given that the rest of the trait solver is sound) over maybe sound but may accept slightly more code.

I've spend a few hours looking at the code and experimenting with some tests to convince myself that your approach is sound (or to get a counter example) and wasn't able to do either of these. Apparently removing the `ocx.register_obligation` doesn't actually break any existing tests but an assertion that the `obligation` has no inference variables does fail. cc @aliemjay as you may know whether that `register_obligation` is currently still needed.

Reply:
> > 
> 
> We do care about them if these other predicates end up causing us to add additional outlives predicates. Considering that we end up computing `Projection` obligations inside of `compute_implied_outlives_bounds`, this can be the case.

In `implied_outlives_bounds` query, `ocx.register_obligations` will add additional predicates as requirements not assumptions. i.e. the additional outlives predicates will end up in `QueryResponse::region_constraints` rather than `QueryResponse::value`.

> Apparently removing the ocx.register_obligation doesn't actually break any existing tests but an assertion that the obligation has no inference variables does fail. cc @aliemjay as you may know whether that register_obligation is currently still needed.

After fiddling with it for a while I managed to come up with a test:
```rust
pub trait Iter {
    type Item;
}

impl<X, I> Iter for I
where
    I: IntoIterator<Item = X>,
{
    type Item = X;
}

pub struct Map<I>(I)
where
    I: Iter,
    I::Item: 'static;

fn test<X>(_: Map<Vec<X>>) {}
```
and filed an issue #109799 for why the condition `predicate.has_non_region_infer()` is unnecessary.

Reply:
Extending the approach in https://github.com/rust-lang/rust/pull/109491 ended up causing weird trait solver errors so I changed my opinion here. I think that - at least with the current solver - this PR is the better choice.

Please add the following as a test.
```rust
// check-pass

// Similar to issue-108544.rs except that we have a generic `T` which
// previously caused an overeager fast-path to trigger.
use std::borrow::Cow;

pub trait Trait<T: Clone> {
    fn method(self) -> Option<Cow<'static, T>>
    where
        Self: Sized;
}

impl<'a, T: Clone> Trait<T> for Cow<'a, T> {
    fn method(self) -> Option<Cow<'static, T>> {
        None
    }
}

fn main() {}
```